synapse/metrics/background_process_metrics.py

```diff
@@ -169,7 +188,19 @@ def update_metrics(self):
         )
 
 
-def run_as_background_process(desc: str, func, *args, bg_start_span=True, **kwargs):
+R = TypeVar("R")
+
+
+def run_as_background_process(
+    desc: str,
+    func: Union[
+        Callable[..., Awaitable[R]],
+        # Callable[..., R],
```

Comment:
Uncommenting this results in
```
synapse/util/distributor.py:74: error: Argument 2 to "run_as_background_process" has incompatible type "Callable[[VarArg(Any), KwArg(Any)], Deferred[List[Any]]]"; expected "Callable[..., Awaitable[<nothing>]]"  [arg-type]
```
and a bunch more errors.

On the other hand, omitting it from the union results in
```
synapse/util/caches/expiringcache.py:86: error: Argument 2 to "run_as_background_process" has incompatible type "Callable[[], None]"; expected "Callable[..., Awaitable[Any]]"  [arg-type]
```

Reply:
Fixed by using overloads and suppressing the error about overlapping signatures.